Add system.query_metric_log

pamarcos opened the pull request:
Add `system.query_metric_log` which contains history of memory and metric values from table system.events for individual queries, periodically flushed to disk.

The interval can be configured on a per-query basis through the new `query_metric_log` setting. For `ProfileEvents` we collect all of them per query. For `CurrentMetrics` we only collect two memory-related metrics because they're useful, but the rest don't make much sense, since they're global metrics anyway. Take a look at this [thread for more context](https://clickhouse-inc.slack.com/archives/C02F2LML5UG/p1720188405899809?thread_ts=1720174804.766799&cid=C02F2LML5UG).

Since we use [getrusage](https://www.man7.org/linux/man-pages/man2/getrusage.2.html) to collect profile counters, we can only update them from the same thread the query is running. Initially I hooked into the `progressCallback` for that purpose to call it more often. However, `ReadProgressCallback` is already calling it often enough.

For collecting the data, we can't use a periodic background thread the same way we do for `metric_log` or `error_log`. The difference is that for `query_metric_log` we're collecting at periodic intervals **for each individual query** after a query has started and also when the query finishes (if the query takes longer than the collection interval). We use the common `BackgroundSchedulePool` of threads so the tasks to collect the info are worked on. In other words, we're not collecting all queries at the same periodic timestamp because they can be tuned per query. This means, a query A starting at timestamp 0 and query B starting at 100 with default interval of 1000ms can write rows as follows:

```
timestamp 0: nothing, since query A just started and there's no prior profile counters to compare against
timestamp 100: nothing, since query B just started and there's no prior profile counters to compare against
timestamp 1000: row for query A
timestamp 1100: row for query B
timestamp 2000: row for query A
timestamp 2100: row for query B
timestamp 2567: query A finishes, new row
timestamp 2890: query B finishes, new row
```

Regarding the new test introduced, I've executed them 5000 times concurrently with no flakiness:

```bash
tests/clickhouse-test 03203_system_query_metric_log --test-run=5000 -j32
``` 

I've made a test with the dashboard showing some metrics from the [UK property prices dataset
](https://clickhouse.com/docs/en/getting-started/example-datasets/uk-price-paid) and looks quite nice. I haven't done any aggregation on purpose to see exactly what the individual rows of the table contain.

![image](https://github.com/user-attachments/assets/a293ccab-df1d-4dbb-8178-b22a3ce932e5)

The dashboard:

```JSON
{
  "params": {
    "query_id": "bb369149-866a-4cf3-b3e2-cb8debd8f6af"
  },
  "queries": [
    {
      "title": "Memory usage",
      "query": "select event_time_microseconds::Int as t,\nmemory_usage\nfrom system.query_metric_log\nwhere query_id = {query_id:String}\norder by t"
    },
    {
      "title": "Peak Memory usage",
      "query": "select event_time_microseconds::Int as t,\npeak_memory_usage\nfrom system.query_metric_log\nwhere query_id = {query_id:String}\norder by t"
    },
    {
      "title": "Inserted rows",
      "query": "select event_time_microseconds::Int as t,\npeak_memory_usage\nfrom system.query_metric_log\nwhere query_id = {query_id:String}\norder by t"
    },
    {
      "title": "Inserted bytes",
      "query": "select event_time_microseconds::Int as t,\nProfileEvent_InsertedBytes\nfrom system.query_metric_log\nwhere query_id = {query_id:String}\norder by t"
    },
    {
      "title": "SelectedBytes",
      "query": "select event_time_microseconds::Int as t, ProfileEvent_InsertedWideParts\nfrom system.query_metric_log\nwhere query_id = {query_id:String}\norder by t"
    },
    {
      "title": "ReadCompressedBytes",
      "query": "select event_time_microseconds::Int as t, ProfileEvent_ReadCompressedBytes\nfrom system.query_metric_log\nwhere query_id = {query_id:String}\norder by t"
    },
    {
      "title": "InsertedWideParts",
      "query": "select event_time_microseconds::Int as t, ProfileEvent_InsertedWideParts\nfrom system.query_metric_log\nwhere query_id = {query_id:String}\norder by t"
    },
    {
      "title": "NetworkReceiveElapsedMicroseconds",
      "query": "select event_time_microseconds::Int as t, ProfileEvent_NetworkReceiveElapsedMicroseconds\nfrom system.query_metric_log\nwhere query_id = {query_id:String}\norder by t"
    }
  ]
}
```

Closes #40012 

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add system.query_metric_log which contains history of memory and metric values from table system.events for individual queries, periodically flushed to disk.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [x] <!---no_ci_cache--> Disable CI cache

pamarcos committed:
Add system.query_log_metric table

pamarcos committed:
Update query_log_metric using the progress callback

pamarcos committed:
Move QueryLogMetric to be a periodic log

pamarcos committed:
Collect system.query_log_metric at periodic times

pamarcos committed:
Add naïve implementation of per-query metric

pamarcos committed:
Use the proper data structure for queries

pamarcos committed:
Add per-query tracking with different intervals

pamarcos committed:
Add test and simplify logic

pamarcos committed:
Fix exception text with proper range

pamarcos committed:
Add lots of debugging traces to figure out what's happening

pamarcos committed:
Use boost::multi_index_container instead of std::set

This way we avoid having to create a new container within
stepFunction, since the multi-index set allows us to modify
the values while having several indexes without worrying
about the underneath structure(s).

pamarcos committed:
Update test to make it more stable

Let's use 5s instead of 3s to have less possibilities of ending
up with just 1 diff value for 1000ms interval.

Also, let's remove the logging traces used to debug.

pamarcos committed:
Fix regex to match only specific tables

Previously was matching system.query_log_metric because
system.query_log is contained within.

pamarcos committed:
Fix check-style

pamarcos committed:
Make getting iterators more user friendly

pamarcos committed:
Add a 11ms concurrent query

pamarcos committed:
Add documentation for system.query_log_metric

robot-ch-test-poll commented:
<!-- automatic status comment for PR #66532 from pamarcos/ClickHouse:system-query-metric-log -->
*This is an automated comment for commit 10945fe108a6c332479621c25ea8743dc2934a13 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66532/10945fe108a6c332479621c25ea8743dc2934a13/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66532/10945fe108a6c332479621c25ea8743dc2934a13/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

pamarcos committed:
Update documentation

pamarcos committed:
Normalize tag name to meet ClickHouse convention instead of boost

pamarcos committed:
Fix previous commit

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Fix segfault whenever there's no context

pamarcos committed:
Add query_log_metric to config samples

pamarcos committed:
Add new setting to SettingsChangesHistory

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Relax the test condition

Let's use a slower interval than default 1000ms and
relax the stddev condition to make it less flaky.

pamarcos committed:
Fix doc

pamarcos commented on src/Common/SystemLogBase.cpp:
```diff
@@ -262,8 +264,10 @@ void SystemLogBase<LogElement>::notifyFlush(bool force) { queue->notifyFlush(for
 
 #define INSTANTIATE_SYSTEM_LOG_BASE(ELEMENT) template class SystemLogBase<ELEMENT>;
 SYSTEM_LOG_ELEMENTS(INSTANTIATE_SYSTEM_LOG_BASE)
+SYSTEM_PERIODIC_LOG_ELEMENTS(INSTANTIATE_SYSTEM_LOG_BASE)
```
I moved this here to avoid adding the periodic system log names both to `SYSTEM_LOG_ELEMENTS` and `SYSTEM_PERIODIC_LOG_ELEMENTS`. This way, there's only one define.

pamarcos committed:
Fix silly mistake collecting ProfileEvents

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Fix configs for query_log_metric

pamarcos committed:
Remove redundant metric_log, error_log and query_log_metric conf

They're already included as separate files. There's no need to have
them in the main config.

pamarcos committed:
Clarify comment

pamarcos committed:
Add missing query_log_metric config

pamarcos committed:
Rename system.query_log_metric to system.query_metric_log

Seems I had at the very beginning a brainfart and started using
query_log_metric instead of query_metric_log and it went on and on
:facepalm

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Set thread names for all Periodic Logs

pamarcos committed:
Fix shutdown process :)

pamarcos committed:
Make sure there are no spurious wakeups

pamarcos committed:
Add logical error for unexpected behavior

pamarcos committed:
Remove the 47ms interval query because it's flaky

Even though running 5000 times the test in parallel
with 32 jobs in my dev machine doesn't show any error,
the CI does complain about the stddev of this one.
So, rather than increasing the range for all values
let's remove this one because for time-critical things
depending on the resources the result may be flaky.

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Fix SettingsChangesHistory after merge

pamarcos committed:
Remove context from QueryMetricLog since it was already there

At the expense of inheriting publicly from WeakContext
in SystemLog.

pamarcos committed:
Simplify logic using BackgroundSchedulePool

- Use the common schedule pool to collect the data, instead
of having a somewhat duplicate logic to do the same in a
single thread. This is better for code maintenance and also
for performance, since now there's a myriad of threads where
any of them may do the collection. Of course, I did not know
of the existence of the schedule pool before :)

- Move away from PeriodicLog for QueryMetricLog since in
doesn't make sense anymore.

pamarcos committed:
Add logical error for unexpected behavior

pamarcos committed:
Fix finishQuery cause can be called without a query started

pamarcos committed:
Ensure query_metric_log is stopped during shutdown

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Fix after the automatic merge conflict resolution

pamarcos committed:
Fix leftover after merge

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Merge branch 'master' into system-query-metric-log

Algunenano commented on docs/en/operations/system-tables/query_metric_log.md:
```diff
@@ -0,0 +1,48 @@
+event_date:                                                      2020-09-05
+event_time:                                                      2020-09-05 16:22:33
+event_time_microseconds:                                         2020-09-05 16:22:33.196807
+CurrentMetric_MemoryTracking:                                    480794407
```
Can we show `query_id` before other metrics (and include it in the example)?

Algunenano commented on src/Core/SettingsChangesHistory.cpp:
```diff
@@ -81,6 +81,7 @@ static std::initializer_list<std::pair<ClickHouseVersion, SettingsChangesHistory
             {"restore_replace_external_engines_to_null", false, false, "New setting."},
             {"input_format_json_max_depth", 1000000, 1000, "It was unlimited in previous versions, but that was unsafe."},
             {"merge_tree_min_bytes_per_task_for_remote_reading", 4194304, 2097152, "Value is unified with `filesystem_prefetch_min_bytes_for_single_read_task`"},
+            {"query_metric_log_interval", 0, 0, "New setting."},
```
Please move it to 24.9

Algunenano commented on src/Interpreters/QueryMetricLog.cpp:
```diff
@@ -0,0 +1,180 @@
+
+    auto context = getContext();
+    const auto & process_list = context->getProcessList();
+    status.task = context->getSchedulePool().createTask("QueryLog", [this, &process_list, query_id] {
```
So essentially we are adding a new thread per query that we want to monitor. I don't know how I feel about that. If you have 200 concurrent queries you'll add an extra 200 threads if you have system.query_metric_log enabled :S

Algunenano replied:
At least we should be really clear about the impact this will have in terms of extra resource usage per query

pamarcos replied:
>So essentially we are adding a new thread per query that we want to monitor.

If I understood correctly the `BackgroundSchedulePool` logic, that's not the case. `QueryMetricLog` is creating a task that is scheduled to be run by one of the threads available in this common pool. It is not spawning a new thread each query. That'd be indeed not such a great idea and would have more implications as you suggested.

Some context: prior to [this commit](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/befb90b7ca1598b6da3da915e7b85367fae6f3bc) the `QueryMetricLog` was spawning its own thread that had a very similar logic to what `BackgroundSchedulePool` has, with similar condition variables, mutexes and all that. Then, I saw `BackgroundSchedulePool` and realized we'd rather use that implementation instead.

Algunenano commented on src/Interpreters/QueryMetricLog.cpp:
```diff
@@ -0,0 +1,180 @@
+    elem.event_time_microseconds = timeInMicroseconds(current_time);
+    elem.query_id = query_status_it->first;
+    elem.memory = CurrentMetrics::values[CurrentMetrics::MemoryTracking];
+    elem.background_memory = CurrentMetrics::values[CurrentMetrics::MergesMutationsMemoryTracking];
```
This seems odd. Are this values global (which wouldn't make sense to log as part of query metrics) or per thread (which then could not have `background_memory`) or per query somehow?

Related to this: I would expect to see `memory_usage` and `peak_memory_usage` from `QueryStatusInfo` as we do for query_log. Is there an issue with those values that leads us to use CurrentMetrics instead?

pamarcos replied:
This was indeed wrong. The `memory_usage` for the query and the `CurrentMetrics::MemoryTracking` was the same on the first implementation of this where I was collecting the data **on the same thread the query runs** (in the progressCallback hook). However, this is not the case anymore, so I was taking the memory value for the pool thread, which is very off.

Agree with you, I'm going to report just the `memory_usage` and `peak_memory_usage`. The former will be now the correct value for the query, and the `background_memory` doesn't make much sense.

Algunenano commented on src/Interpreters/QueryMetricLog.h:
```diff
@@ -0,0 +1,66 @@
+
+#include <Common/ProfileEvents.h>
+#include <Common/CurrentMetrics.h>
+#include "Core/BackgroundSchedulePool.h"
```
```suggestion
#include <Core/BackgroundSchedulePool.h>
```

pamarcos replied:
Yep yep, that's VSCode adding headers automatically. I missed this one. I can't see how to configure it to use <> instead of double quotes.

Algunenano commented on src/Interpreters/QueryMetricLog.cpp:
```diff
@@ -0,0 +1,180 @@
+                parseQuery(codec_parser, "(Delta(4), ZSTD(1))", 0, DBMS_DEFAULT_MAX_PARSER_DEPTH, DBMS_DEFAULT_MAX_PARSER_BACKTRACKS),
+                "Event time."});
+    result.add({"event_time_microseconds",
+                std::make_shared<DataTypeDateTime64>(6),
```
Why doesn't this column has a codec and the others do? Since event_time_microseconds will increase in a similar fashion as event_time it'd make sense to have a similar codec. OTOH, I'm not sure if codecs are useful at all here, and it might be that ZSTD by itself is more than enough

pamarcos replied:
The answer is that no other `event_time_microseconds` for any other log had any codec previously. The rest of the rows here have codecs since I created [ErrorLog](https://github.com/ClickHouse/ClickHouse/issues/64501), but prior to that one none had codecs.

I'll add the `Delta(4), ZSTD(1)` codec for consistency, in any case :+1:

Algunenano commented on src/Interpreters/QueryMetricLog.cpp:
```diff
@@ -0,0 +1,180 @@
+
+    for (const auto & metric : memory_metrics)
+    {
+        auto name = fmt::format("CurrentMetric_{}", CurrentMetrics::getName(metric));
```
If we are only going to save 2 metrics related to memory, maybe we can simply forget about CurrentMetrics and save the values in the element itself. WDYT?

Algunenano commented on src/Processors/Transforms/WindowTransform.cpp:
```diff
@@ -2480,7 +2480,7 @@ struct WindowFunctionNthValue final : public WindowFunction
         {
             throw Exception(ErrorCodes::BAD_ARGUMENTS,
-                "The offset for function {} must be in (0, {}], {} given",
+                "The offset for function {} must be in (1, {}], {} given",
```
Seems unrelated

pamarcos replied:
It is indeed unrelated. Is something I found along the way and I piggybacked it on this PR. Here's the commit: https://github.com/ClickHouse/ClickHouse/pull/66532/commits/0557ef17304eaab590485c4d4038b7eef61ce920
Should I open a separate PR just for this?

Algunenano commented on src/QueryPipeline/QueryPipeline.cpp:
```diff
@@ -572,7 +571,25 @@ Block QueryPipeline::getHeader() const
+        if (context)
+        {
+            if (auto query_metric_log = context->getQueryMetricLog())
+                CurrentThread::updatePerformanceCountersIfNeeded();
```
I'm guessing this will be quite inexact, but not more than the current counters we already have for queries and so on.

pamarcos replied:
That's also my assumption

Algunenano reviewed (commented):
Pretty nice. I've left some comments.

An idea to verify usefulness an impact would be to add this to the CI of stateless tests and save it, similar to what we do with trace logs. That way we could test a) an intensive scenario and b) whether we can use it for investigate slow tests (of which we have many)

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Move query_id to top

pamarcos committed:
Refactor to make it easier to reason about

pamarcos committed:
Fix task log name

pamarcos commented on docs/en/operations/settings/settings.md:
```diff
@@ -1831,6 +1831,13 @@ Possible values:
+
+The interval in milliseconds at which the [query_metric_log](../../operations/system-tables/query_metric_log.md) for individual queries is collected.
+If set to 0, it will take the `collect_interval_milliseconds` from the [query_metric_log setting](../../operations/server-configuration-parameters/settings.md#query_metric_log).
+
```
Fair enough. I'm going to change the default value to be -1 meaning it'd take the default value at `collect_interval_milliseconds`. Then, 0 means we want to disable the collection for that particular query. Sounds good?

pamarcos committed:
Disable query_metric_log collection setting the interval to 0

pamarcos committed:
Remove the `CurrentMetric_` prefix

pamarcos committed:
Add codec for `event_time_microseconds`

pamarcos commented:
> An idea to verify usefulness an impact would be to add this to the CI of stateless tests and save it, similar to what we do with trace logs. That way we could test a) an intensive scenario and b) whether we can use it for investigate slow tests (of which we have many)

Those tests are already using this since I modified the [programs/server/config.xml](https://github.com/ClickHouse/ClickHouse/pull/66532/files#diff-8a65ed88620ac5ef2253eb77a159f9b788d13860393c44db27a1c45846e3e7f0). However, it's not easily accessible in as an asset for the CI. Let me take a look how to store and present it in a similar way to how it's done for `query_log` and others.

pamarcos committed:
Save system.query_metric_log as a CI artifact

pamarcos committed:
Go back to a more relaxed check

On the CI it has failed in the case of the smallest interval

pamarcos committed:
Add missing system.query_metric_log artifact

pamarcos committed:
Fix memory collection

Use the memory_usage and peak_memory_usage from the QueryStatusInfo
instead which holds the proper value independently of the thread
we are collecting the data from.

pamarcos committed:
Remove unnecessary externs

pamarcos committed:
Make tests faster and more deterministic

Running the queries for intervals that are not a multiple
of the query_metric_log_intervals ensures there's no
collection at the edge.

pamarcos committed:
Collect on query finish if took longer than the interval

pamarcos committed:
Defensive programming to avoid leaking resources

pamarcos committed:
Revert "Defensive programming to avoid leaking resources"

This reverts commit 181bc7f34c1c84c37b06967f7406fc366319a355.

pamarcos committed:
Move BackgroundSchedulePool::TaskHolder's ownership to QueryStatus

This way we ensure that the background task is encapsulated within
QueryStatus, ensuring we will never ever leak resources once the
query finishes.

Code-wise the previous implementation was ok, but this is making
it more bullet-proof for future changes.

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Relax the logical errors because they could happen on very fast queries

pamarcos committed:
Remove unnecessary update of performance counters

They're already updated enough in the ReadProgressCallback

pamarcos committed:
Relax another logical error that could happen

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Fix segmentation fault in case profile events is not collected

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Refactor to simplify when and how to call finishQuery

pamarcos committed:
Add a separate background schedule pool for QueryMetricLog

This prevents deadlocks and more funny stuff happening in case
the shared background schedule pool ends up executing a query
that will eventually be logged in another task within that
same pool.

pamarcos committed:
Remove server setting from changes history

pamarcos committed:
Change default number of threads to collect QueryMetricLog

pamarcos committed:
Fix deadlock :)

pamarcos commented:
I've reproduced the deadlock with the following minimum example, while staying very close to the original issue:

```C++
#include <semaphore>
#include <list>
#include <iostream>

#include <Common/Logger.h>
#include <Common/logger_useful.h>
#include <Core/BackgroundSchedulePool.h>
#include <Interpreters/Context.h>
#include <Poco/ConsoleChannel.h>
#include <Poco/Logger.h>
#include <Poco/Process.h>
#include <Poco/Thread.h>

using namespace DB;

namespace CurrentMetrics
{
    extern const Metric BackgroundQueryMetricLogSchedulePoolTask;
    extern const Metric BackgroundQueryMetricLogSchedulePoolSize;
}

struct MyQueryStatus
{
    explicit MyQueryStatus(std::unique_ptr<BackgroundSchedulePool::TaskHolder> task) :
        query_metric_log_task(std::move(task))
    {}

    std::unique_ptr<BackgroundSchedulePool::TaskHolder> query_metric_log_task;
};

using MyQueryStatusPtr = std::shared_ptr<MyQueryStatus>;

std::list<MyQueryStatusPtr> processes;
LoggerPtr logger;

void getInfo()
{
    LOG_DEBUG(logger, "getInfo");
    auto process = processes.front();

    /// Commenting out the next line prevents the deadlock, since process doesn't
    /// hold anymore the last reference
    processes.clear();
}

int main()
{
    auto shared_context = Context::createShared();
    auto global_context = Context::createGlobal(shared_context.get());
    global_context->makeGlobalContext();

    Poco::AutoPtr<Poco::ConsoleChannel> channel(new Poco::ConsoleChannel(std::cerr));
    Poco::Logger::root().setChannel(channel);
    Poco::Logger::root().setLevel("debug");
    logger = getLogger("PMO");
    std::binary_semaphore semaphore(0);

    LOG_DEBUG(logger, "Hello!");

    BackgroundSchedulePool pool(
        2,
        CurrentMetrics::BackgroundQueryMetricLogSchedulePoolTask,
        CurrentMetrics::BackgroundQueryMetricLogSchedulePoolSize,
        "BgQMLSchPool");

    auto task = std::make_unique<BackgroundSchedulePool::TaskHolder>(pool.createTask("my_task", [&semaphore]() {
        LOG_DEBUG(logger, "Starting task...");
        getInfo();
        LOG_DEBUG(logger, "Finished task...");
        semaphore.release();
    }));
    (*task)->schedule();

    processes.emplace_back(std::make_shared<MyQueryStatus>(std::move(task)));
    semaphore.acquire();

    return 0;
}
```

Leaving it here for reference and in case it helps understanding the reasoning behind 9feeed70b6473780c18734aba9dd99b07f50b34c

Algunenano commented on src/Interpreters/ProcessList.h:
```diff
@@ -400,6 +400,7 @@ class ProcessList : public ProcessListBase
     /// Call under lock. Finds process with specified current_user and current_query_id.
     QueryStatusPtr tryGetProcessListElement(const String & current_query_id, const String & current_user);
 
+    /// Call under lock. Finds process with specified query_id.
```
Can we use TSA annotations instead? They are much better than just comments

pamarcos replied:
Done in [`e1e2d07` (#66532)](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/e1e2d07341afd30bbefcd9babc74f2832a2dc049)

Took much longer than I anticipated :sweat:

pamarcos committed:
Add trace logs to figure out what's happening on the CI

pamarcos committed:
Refactor to allow TSA annotations

There doen't seem to be any way to make safeLock() or unsafeLock()
work with TSA annotations because the analyzer always thinks that
the mutex is hold and never released in the temporary var of the
body implementation.

Thus, I had to refactor to construct directly the proper locks
wherever they where needed:

safeLock() -> LockAndBlock
unsafeLock() -> Lock

Lastly, in those places where a std::unique_lock was needed because
it's the interface used for conditional_variables and others,
we use that instead of the annotated LockGuard.

pamarcos committed:
WIP

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Revert "WIP"

This reverts commit e1f40bc4d936320a7462f768e4379856e1613c94.

pamarcos committed:
Update submodules

pamarcos committed:
Fix merge messup

pamarcos committed:
Fix style check

pamarcos committed:
Avoid scheduling next query log metric task if not needed

pamarcos committed:
Fix typo

pamarcos committed:
Remove some log traces

pamarcos committed:
WIP

pamarcos committed:
Remove more log traces

pamarcos committed:
Move the task holder back to QueryMetricLogStatus

This prevents a race condition in which the destructor of QueryStatus
attempts to lock the exec_lock when deactivating the task. On the other
hand, that lock is hold by the execution of the periodic task itself.

But then, the task is locked because the destructor of QueryStatus is
holding the ProcessList lock. Thus, both threads are blocked on each
other.

pamarcos committed:
Fix style check

pamarcos committed:
Remove unnecessary unique_ptr

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Fix deadlock ^_^

pamarcos committed:
This time is personal

pamarcos committed:
Remove debug traces

pamarcos committed:
Remove more unneeded debug traces

pamarcos committed:
Yet another deadlock fix attempt

pamarcos committed:
Remove unused variable

pamarcos committed:
Relax the threshold for test

pamarcos committed:
Clarify comment that fixes deadlock

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Revert "Add a separate background schedule pool for QueryMetricLog"

This reverts commit 03685baec65139afd269fd30237a1758bcb8b8ce.

Now that the deadlock is properly fixed, there's no real need to have
a separate thread pool for QueryMetricLog.

pamarcos committed:
WIP

pamarcos committed:
Add debug and trace logs to help figuring out issues in the future

pamarcos committed:
Revert "WIP"

This reverts commit 3c5bdc7cd90d38070865e14071ac41fee46e657c.

pamarcos commented:
@Algunenano ready for another round of review :smile: 
No more deadlocks :rocket:

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Remove stddev because the CI is far from deterministic

The CI is often times overloaded and the background threads
that collect the query metric logs take more time to pick up
the tasks.

So, let's stick to the average of the diff for the sake of
avoiding flakiness.

pamarcos commented:
I ended up removing the low stddev check in the test at [`532d195` (#66532)](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/532d1951b75deac25a9750e4414dd0777e2859e6) because it happened on the CI that whenever there are lots of queries processed, the task is not scheduled for a while and then suddenly all of them are scheduled. It could be because the background schedule pool is overloaded, but taking into account there are 512 threads for that, I'm more inclined to blame the `queries_mutex` and the mutexes of the background schedule pool used to schedule tasks.

# query_metric_log

```sql
WITH diff AS                                                                                                                                                    
    (                                                                                                                                                           
        SELECT                                                                                                                                                  
            row_number() OVER () AS row,                                                                                                                        
            count() OVER () AS total_rows,                                                                                                                      
            event_time_microseconds,                                                                                                                            
            first_value(event_time_microseconds) OVER (ORDER BY event_time_microseconds ASC ROWS BETWEEN 1 PRECEDING AND 0 FOLLOWING) AS prev,                  
            dateDiff('ms', prev, event_time_microseconds) AS diff                                                                                               
        FROM file('query_metric_log.tsv')                                                                                                                       
        WHERE query_id = 'test_2ehnkw4z_123'                                                                                                                    
        ORDER BY event_time_microseconds ASC                                                                                                                    
        OFFSET 1                                                                                                                                                
    )                                                                                                                                                           
SELECT *                                                                        
FROM diff                                                                                                                                                       
                                                                                                                                                                
Query id: e88b2d08-8c77-431b-a8c5-c453f3adc737                                                                                                                  
                                                                                                                                                                
    ┌─row─┬─total_rows─┬────event_time_microseconds─┬───────────────────────prev─┬─diff─┐                                                                       
 1. │   2 │         20 │ 2024-10-17 14:39:10.617758 │ 2024-10-17 14:39:10.502972 │  115 │                                                                       
 2. │   3 │         20 │ 2024-10-17 14:39:10.735886 │ 2024-10-17 14:39:10.617758 │  118 │                                                                       
 3. │   4 │         20 │ 2024-10-17 14:39:10.859897 │ 2024-10-17 14:39:10.735886 │  124 │                                                                       
 4. │   5 │         20 │ 2024-10-17 14:39:10.982036 │ 2024-10-17 14:39:10.859897 │  123 │                                                                       
 5. │   6 │         20 │ 2024-10-17 14:39:11.104732 │ 2024-10-17 14:39:10.982036 │  122 │                                                                       
 6. │   7 │         20 │ 2024-10-17 14:39:11.343393 │ 2024-10-17 14:39:11.104732 │  239 │                                                                       
 7. │   8 │         20 │ 2024-10-17 14:39:11.350821 │ 2024-10-17 14:39:11.343393 │    7 │                                                                       
 8. │   9 │         20 │ 2024-10-17 14:39:11.474292 │ 2024-10-17 14:39:11.350821 │  124 │                                                                       
 9. │  10 │         20 │ 2024-10-17 14:39:11.597432 │ 2024-10-17 14:39:11.474292 │  123 │                                                                       
10. │  11 │         20 │ 2024-10-17 14:39:11.720717 │ 2024-10-17 14:39:11.597432 │  123 │                                                                       
11. │  12 │         20 │ 2024-10-17 14:39:11.967130 │ 2024-10-17 14:39:11.720717 │  247 │                                                                       
12. │  13 │         20 │ 2024-10-17 14:39:11.967248 │ 2024-10-17 14:39:11.967130 │    0 │                                                                       
13. │  14 │         20 │ 2024-10-17 14:39:12.089366 │ 2024-10-17 14:39:11.967248 │  122 │                                                                       
14. │  15 │         20 │ 2024-10-17 14:39:12.212680 │ 2024-10-17 14:39:12.089366 │  123 │                                                                       
15. │  16 │         20 │ 2024-10-17 14:39:12.335106 │ 2024-10-17 14:39:12.212680 │  123 │                                                                       
16. │  17 │         20 │ 2024-10-17 14:39:12.458097 │ 2024-10-17 14:39:12.335106 │  123 │                                                                       
17. │  18 │         20 │ 2024-10-17 14:39:12.581448 │ 2024-10-17 14:39:12.458097 │  123 │                                                                       
18. │  19 │         20 │ 2024-10-17 14:39:13.107684 │ 2024-10-17 14:39:12.581448 │  526 │                                                                       
19. │  20 │         20 │ 2024-10-17 14:39:13.108352 │ 2024-10-17 14:39:13.107684 │    1 │                                                                       
    └─────┴────────────┴────────────────────────────┴────────────────────────────┴──────┘                                                                       

19 rows in set. Elapsed: 0.140 sec. Processed 6.01 thousand rows, 10.49 MB (42.98 thousand rows/s., 74.98 MB/s.)
Peak memory usage: 58.61 MiB. 
```

# query_log

```sql
SELECT                               
    toStartOfMinute(event_time_microseconds) AS time,                                     
    count()                          
FROM file('query_log.tsv')           
WHERE type > 1                       
GROUP BY time                        
ORDER BY time ASC                    
FORMAT PrettyCompactMonoBlock        
                                        
Query id: 12015387-58bf-4826-b731-5e85cd0ce6db                                            
                                        
    ┌────────────────time─┬─count()─┐
 1. │ 2024-10-17 14:16:00 │     890 │
 2. │ 2024-10-17 14:17:00 │   16736 │                                                                                                                           
 3. │ 2024-10-17 14:18:00 │   16671 │
 4. │ 2024-10-17 14:19:00 │   11966 │
 5. │ 2024-10-17 14:20:00 │    6181 │
 6. │ 2024-10-17 14:21:00 │    6672 │              
 7. │ 2024-10-17 14:22:00 │    4688 │              
 8. │ 2024-10-17 14:23:00 │    6172 │              
 9. │ 2024-10-17 14:24:00 │   17451 │              
10. │ 2024-10-17 14:25:00 │    9332 │              
11. │ 2024-10-17 14:26:00 │   16061 │              
12. │ 2024-10-17 14:27:00 │   17564 │                            
13. │ 2024-10-17 14:28:00 │   11917 │                                                                                                                                                                        
14. │ 2024-10-17 14:29:00 │    8794 │              
15. │ 2024-10-17 14:30:00 │   12690 │                            
16. │ 2024-10-17 14:31:00 │    3930 │              
17. │ 2024-10-17 14:32:00 │    5137 │                            
18. │ 2024-10-17 14:33:00 │   12694 │                            
19. │ 2024-10-17 14:34:00 │    7655 │                            
20. │ 2024-10-17 14:35:00 │    7567 │                            
21. │ 2024-10-17 14:36:00 │    4497 │                            
22. │ 2024-10-17 14:37:00 │    6476 │                            
23. │ 2024-10-17 14:38:00 │    5655 │                            
24. │ 2024-10-17 14:39:00 │    5088 │                            
25. │ 2024-10-17 14:40:00 │    4201 │                            
26. │ 2024-10-17 14:41:00 │    1304 │                            
27. │ 2024-10-17 14:42:00 │    1590 │                            
28. │ 2024-10-17 14:43:00 │    1405 │                                                     
29. │ 2024-10-17 14:44:00 │    1591 │                                                                                                                                                                                                                                
30. │ 2024-10-17 14:45:00 │   12902 │                            
31. │ 2024-10-17 14:46:00 │     853 │                                                     
32. │ 2024-10-17 14:47:00 │    1345 │                            
33. │ 2024-10-17 14:48:00 │     415 │                                                     
34. │ 2024-10-17 14:49:00 │    2588 │                                                     
35. │ 2024-10-17 14:50:00 │    1494 │                                                     
36. │ 2024-10-17 14:51:00 │    1020 │                                                     
37. │ 2024-10-17 14:52:00 │    2316 │                                                     
38. │ 2024-10-17 14:53:00 │     378 │                                                     
39. │ 2024-10-17 14:54:00 │    5357 │                                                     
40. │ 2024-10-17 14:55:00 │    2868 │                                                     
41. │ 2024-10-17 14:56:00 │    2064 │                                                     
42. │ 2024-10-17 14:57:00 │     200 │                                                     
43. │ 2024-10-17 14:58:00 │     906 │                                                     
44. │ 2024-10-17 14:59:00 │    2343 │                                                     
45. │ 2024-10-17 15:00:00 │     666 │                                                     
46. │ 2024-10-17 15:01:00 │     524 │                                                     
47. │ 2024-10-17 15:02:00 │    1363 │                                                     
48. │ 2024-10-17 15:03:00 │     927 │                                                     
49. │ 2024-10-17 15:04:00 │     166 │                                                     
50. │ 2024-10-17 15:05:00 │    1124 │                                                     
51. │ 2024-10-17 15:06:00 │     597 │                                                     
52. │ 2024-10-17 15:07:00 │     584 │                                                     
53. │ 2024-10-17 15:08:00 │    4264 │                                                     
54. │ 2024-10-17 15:09:00 │    9634 │                                                     
55. │ 2024-10-17 15:10:00 │    2153 │                                                     
56. │ 2024-10-17 15:11:00 │    1461 │                                                     
57. │ 2024-10-17 15:12:00 │     452 │                                                     
58. │ 2024-10-17 15:13:00 │      75 │                                                     
    └────────────────time─┴─count()─┘                                                     

58 rows in set. Elapsed: 5.188 sec. Processed 574.72 thousand rows, 2.73 GB (110.78 thousand rows/s., 525.60 MB/s.)                                                                 
Peak memory usage: 337.89 MiB.                                                            
```

# server log
```
4146087:2024.10.17 14:39:10.503008 [ 1077 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4146476:2024.10.17 14:39:10.617777 [ 872 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4146560:2024.10.17 14:39:10.735911 [ 952 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4147055:2024.10.17 14:39:10.859932 [ 642 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4147560:2024.10.17 14:39:10.982069 [ 1132 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4147649:2024.10.17 14:39:11.104776 [ 772 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4147669:2024.10.17 14:39:11.343707 [ 772 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4147706:2024.10.17 14:39:11.350846 [ 752 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4148834:2024.10.17 14:39:11.474311 [ 1100 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4149732:2024.10.17 14:39:11.597472 [ 1029 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4149843:2024.10.17 14:39:11.967002 [ 1054 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true <---- this
4149852:2024.10.17 14:39:11.967171 [ 946 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true <---- this
4149856:2024.10.17 14:39:11.967261 [ 1010 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true <---- this
4150793:2024.10.17 14:39:12.089395 [ 744 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4151072:2024.10.17 14:39:12.212721 [ 1142 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4151442:2024.10.17 14:39:12.335135 [ 948 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4151557:2024.10.17 14:39:12.459325 [ 846 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4152116:2024.10.17 14:39:12.581489 [ 1059 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4152170:2024.10.17 14:39:13.108255 [ 1104 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: true
4152174:2024.10.17 14:39:13.108360 [ 35741 ] {test_2ehnkw4z_123} <Debug> QueryMetricLog: Collecting query_metric_log for query test_2ehnkw4z_123. Schedule next: false
```

pamarcos committed:
Fix reference

pamarcos committed:
Merge branch 'master' into system-query-metric-log

Algunenano commented on src/QueryPipeline/QueryPipeline.cpp:
```diff
@@ -26,7 +26,6 @@
 #include <Processors/Transforms/StreamInQueryCacheTransform.h>
 #include <Processors/Transforms/TotalsHavingTransform.h>
 #include <QueryPipeline/Chain.h>
-#include <QueryPipeline/Pipe.h>
```
Pipe.h is clearly being used in this file. Whichever linter you are using is wrong

Algunenano commented on tests/queries/0_stateless/03203_system_query_metric_log.sh:
```diff
@@ -0,0 +1,53 @@
+            first_value(event_time_microseconds) OVER (ORDER BY event_time_microseconds ROWS BETWEEN 1 PRECEDING AND 0 FOLLOWING) as prev,
+            dateDiff('ms', prev, event_time_microseconds) AS diff
+        FROM system.query_metric_log
+        WHERE query_id = '${query_prefix}_${interval}'
```
Please filter by event_date too, so this query doesn't take forever on old systems that have been running for a while (`event_date >= yesterday()`)

Algunenano commented on tests/queries/0_stateless/03203_system_query_metric_log.reference:
```diff
@@ -0,0 +1,6 @@
+1	1
```
This reference results are extremely obscure, so when it fails (because it will fail at some point) it will be impossible to understand why.

pamarcos replied:
This should make it easier to understand: https://github.com/pamarcos/ClickHouse/blob/system-query-metric-log/tests/queries/0_stateless/03203_system_query_metric_log.reference

```
number_of_metrics_1000_ok	timestamp_diff_in_metrics_1000_ok
number_of_metrics_1234_ok	timestamp_diff_in_metrics_1234_ok
number_of_metrics_123_ok	timestamp_diff_in_metrics_123_ok
0
0
3
```

Algunenano commented:
Are stress tests and fuzzers using the feature? If not, let's enable it. For stress / upgrade tests we probably can't since the config needs to be the same as for old features, but we can leave it ready to enable it in the next release.

pamarcos commented:
> Are stress tests and fuzzers using the feature? If not, let's enable it. For stress / upgrade tests we probably can't since the config needs to be the same as for old features, but we can leave it ready to enable it in the next release.

Yes, they'll be using it from now on:

- Fuzzer (debug): https://s3.amazonaws.com/clickhouse-test-reports/66532/f002e51a4e196423d7c85f4bd6ca0a68cb35a439/ast_fuzzer__debug_/server.log.zst

```bash
zstdgrep -ina "QueryMetricLog" server.log.zst | head
678:2024.10.18 18:15:03.008009 [ 865 ] {} <Debug> SystemLog (system.query_metric_log): Creating new table system.query_metric_log for QueryMetricLog
2015576:2024.10.18 18:16:56.249907 [ 333 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 2aa5a068-f509-47ec-aa13-7078aaae136b. Schedule next: true
2015757:2024.10.18 18:16:57.249582 [ 237 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 2aa5a068-f509-47ec-aa13-7078aaae136b. Schedule next: true
2015802:2024.10.18 18:16:58.248762 [ 609 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 2aa5a068-f509-47ec-aa13-7078aaae136b. Schedule next: true
2015808:2024.10.18 18:16:59.248954 [ 658 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 2aa5a068-f509-47ec-aa13-7078aaae136b. Schedule next: true
2015810:2024.10.18 18:16:59.429748 [ 191 ] {2aa5a068-f509-47ec-aa13-7078aaae136b} <Debug> QueryMetricLog: Collecting query_metric_log for query 2aa5a068-f509-47ec-aa13-7078aaae136b. Schedule next: false
2015934:2024.10.18 18:17:00.434730 [ 499 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 16f0ce01-1ce5-4419-9353-cbad15e4f342. Schedule next: true
2015935:2024.10.18 18:17:01.433936 [ 416 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 16f0ce01-1ce5-4419-9353-cbad15e4f342. Schedule next: true
2016017:2024.10.18 18:17:02.434118 [ 361 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 16f0ce01-1ce5-4419-9353-cbad15e4f342. Schedule next: true
2016113:2024.10.18 18:17:03.434325 [ 287 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 16f0ce01-1ce5-4419-9353-cbad15e4f342. Schedule next: true
```

- Stress test (debug): https://s3.amazonaws.com/clickhouse-test-reports/66532/f002e51a4e196423d7c85f4bd6ca0a68cb35a439/stress_test__debug_/clickhouse-server.stress.log.zst

```bash
zstdgrep -ina "QueryMetricLog" clickhouse-server.stress.log.zst | head 
3400:2024.10.18 18:19:42.610701 [ 2222 ] {64974dc8-3c2b-4524-a37b-8e3b49cf5d8c} <Debug> QueryMetricLog: Collecting query_metric_log for query 64974dc8-3c2b-4524-a37b-8e3b49cf5d8c. Schedule next: false
3502:2024.10.18 18:19:43.403038 [ 3096 ] {} <Debug> SystemLog (system.query_metric_log): Will use existing table system.query_metric_log for QueryMetricLog
3987:2024.10.18 18:19:46.479381 [ 2339 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9146b200-9f3b-4a0e-af86-127e68eb7eef. Schedule next: true
3989:2024.10.18 18:19:46.487005 [ 2359 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9146b200-9f3b-4a0e-af86-127e68eb7eef. Schedule next: true
4156:2024.10.18 18:19:46.995258 [ 2633 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9146b200-9f3b-4a0e-af86-127e68eb7eef. Schedule next: true
4693:2024.10.18 18:19:48.070445 [ 2704 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9146b200-9f3b-4a0e-af86-127e68eb7eef. Schedule next: true
5185:2024.10.18 18:19:48.996009 [ 2738 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9146b200-9f3b-4a0e-af86-127e68eb7eef. Schedule next: true
5709:2024.10.18 18:19:49.997750 [ 2259 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9146b200-9f3b-4a0e-af86-127e68eb7eef. Schedule next: true
6289:2024.10.18 18:19:50.998830 [ 2724 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9146b200-9f3b-4a0e-af86-127e68eb7eef. Schedule next: true
6473:2024.10.18 18:19:52.013436 [ 2725 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9146b200-9f3b-4a0e-af86-127e68eb7eef. Schedule next: true

```

pamarcos commented on src/Interpreters/QueryMetricLog.cpp:
```diff
@@ -0,0 +1,206 @@
+std::optional<QueryMetricLogElement> QueryMetricLog::createLogMetricElement(const String & query_id, const QueryStatusInfo & query_info, TimePoint current_time, bool schedule_next)
+{
+    std::lock_guard lock(queries_mutex);
+    LOG_DEBUG(logger, "Collecting query_metric_log for query {}. Schedule next: {}", query_id, schedule_next);
```
Yep, agree. I added it after the mutex on purpose to check the timestamp on the CI of where it was really collecting the log to check it was consistent with the table log. I'll move it before the lock.

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos commented on src/Interpreters/ProcessList.h:
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <Core/BackgroundSchedulePool.h>
```
Surprisingly, there was still a reference to something related to that header: https://github.com/pamarcos/ClickHouse/blob/f002e51a4e196423d7c85f4bd6ca0a68cb35a439/src/Interpreters/ProcessList.h#L499-L500

I removed it

pamarcos committed:
Fix issues pointed in review

pamarcos commented on src/Interpreters/QueryMetricLog.cpp:
```diff
@@ -0,0 +1,206 @@
+    status.interval_milliseconds = interval_milliseconds;
+    status.next_collect_time = query_start_time + std::chrono::milliseconds(interval_milliseconds);
+
+    const auto & profile_events = CurrentThread::getProfileEvents();
```
>It seems that whatever events where triggered before calling QueryMetricLog::startQuery won't be part of any log (not even the first for the query), which is kind of unexpected

That's right. `query_metric_log` works exactly the same as `metric_log`, but for particular queries. This means that we store the state at the beginning and then we report periodically the difference between the new state and the previous one. So, as your comment suggests, if there's a single profile event that happens right before this call that is never modified during the execution of the query, it won't be reported in any event of `query_metric_log`. `QueryMetricLog::startQuery` is already called on the early stage of the query execution, but there's stuff that may happen before it for sure. For getting those sort of edge cases we'll need to go to the global `metric_log`.

The best I can think of is moving the call to `QueryMetricLog::startQuery` as early as possible in an attempt to catch every profile event triggered by the query, even at the early stage.

Algunenano replied:
But why ignore it at all for the first instance? I assume profileEvents are 0 at the very start of the query, so they should be there in the first row that dumps them to system.query_metric_log even if they happen before `QueryMetricLog::startQuery`

pamarcos replied:
>I assume profileEvents are 0 at the very start of the query

Yep, I double checked just in case and it's like that: https://github.com/ClickHouse/ClickHouse/blob/4baa30c06d7e7996c548b85120044ca691ff2326/src/Interpreters/ThreadStatusExt.cpp#L383-L385

>But why ignore it at all for the first instance?

I see your point now. It's a decision I made in an attempt to save only the minimum amount of info because I assumed the first row would be of little value. Regarding this, these metrics are only saved once the query exceeds the collection interval (1s by default). In any case, the same way I ended up adding the finish row because I think it's very useful, I can add the start row to catch that case. Will do :+1.

Algunenano replied:
> I see your point now. It's a decision I made in an attempt to save only the minimum amount of info because I assumed the first row would be of little value.

I'm not talking about saving the values at the start of the query (one extra row), but about not discarding the profileEvents that happened before start of query (typically from planning, which can be expensive) so they are dumped when the first row is stored (1 second in, by default). The storage usage should be pretty much the same

pamarcos replied:
Done at [`4223bb9` (#66532)](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/4223bb95a50fde3155ae3d5f02eaf0a95a2e3ec4)

pamarcos replied:
> I'm not talking about saving the values at the start of the query (one extra row), but about not discarding the profileEvents that happened before start of query (typically from planning, which can be expensive) so they are dumped when the first row is stored (1 second in, by default). The storage usage should be pretty much the same

Damn, I misunderstood and implemented the other thing. I reverted it and finally did what you were proposing at [`795abfc` (#66532)](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/795abfcb0a78202967d915bcac707ca9cb6b340c)

pamarcos committed:
Add event at start of query

Also remove the requirement to include events only for queries
that take longer than collection interval.

pamarcos committed:
Revert "Add event at start of query"

This reverts commit 4223bb95a50fde3155ae3d5f02eaf0a95a2e3ec4.

pamarcos committed:
Store every counter before startQuery

Rather than setting the base during startQuery, let the
delta work on every metric collection. The first time,
it'll take 0 as the base and work fine.

Algunenano commented on src/Interpreters/QueryMetricLog.cpp:
```diff
@@ -0,0 +1,201 @@
+    /// The query might have finished while the scheduled task is running.
+    if (query_status_it == queries.end())
+    {
+        LOG_TRACE(logger, "Query {} finished already while this collecting task was running", query_id);
```
This log is still under the mutex :cry:

pamarcos replied:
Done! [`2a1327d` (#66532)](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/2a1327dd91a05f8367e6bf51ae76a03c0ae9f4c7)

Algunenano commented on tests/queries/0_stateless/03203_system_query_metric_log.sh:
```diff
@@ -0,0 +1,55 @@
+{
+    interval=$1
+    # We calculate the diff of each row with its previous row to check whether the intervals at which
+    # data is collected is right. The first row is always skipped because the diff is 0. The same for the
```
`The first row is always skipped because the diff is 0`. Here you mean it's because the preceding row is itself so it will be 0, right? Just confirming

pamarcos replied:
Done at [`4fc4cee` (#66532)](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/4fc4cee8627eae133c18d22fcd297d1445922eb8)

Algunenano commented on tests/queries/0_stateless/03203_system_query_metric_log.sh:
```diff
@@ -0,0 +1,55 @@
+check_log 1000
+check_log 1234
+check_log 123
+
```
It'd be nice to check that we are actually logging something too. Like check that `SleepFunctionCalls` / `SleepFunctionMicroseconds`, `SleepFunctionElapsedMicroseconds` are not 0 (whatever value they must be) in the last / exit log of one of the queries.

pamarcos replied:
Yep, I also thought yesterday about checking `ProfileEvent_Query`, `ProfileEvent_SelectQuery` and `ProfileEvent_InitialQuery` are 1 on the first event to ensure we're not losing anything on the first row.

Will do both :+1:

pamarcos replied:
Done at [`4fc4cee` (#66532)](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/4fc4cee8627eae133c18d22fcd297d1445922eb8)

Algunenano reviewed (approved):
Looks good. Just a couple of minor nitpicks, so feel free to merge once it's green

pamarcos committed:
Move the log trace outside of the mutex

finishQuery doesn't need to take care of this case because
it can never happen here since the query is not removed yet
from the list.

pamarcos committed:
Merge branch 'master' into system-query-metric-log

pamarcos committed:
Improve test to ensure data collected is ok

pamarcos committed:
Fix mistake merging master

azat commented on src/Interpreters/QueryMetricLog.cpp:
```diff
@@ -0,0 +1,200 @@
+
+std::optional<QueryMetricLogElement> QueryMetricLog::createLogMetricElement(const String & query_id, const QueryStatusInfo & query_info, TimePoint current_time, bool schedule_next)
+{
+    LOG_DEBUG(logger, "Collecting query_metric_log for query {}. Schedule next: {}", query_id, schedule_next);
```
Debug logging looks annoying, maybe `LOG_TEST` will be enough?

pamarcos replied:
I'll demote it to `LOG_TRACE` in another PR where I'm working on some improvements for this :+1: 
I'm choosing `trace` instead of `test` because that's the default level for the CI and it's helped me a lot to debug issues only seen on the CI

azat replied:
After giving it another thought `LOG_TRACE` still can be too much, especially due to it is the default and what in general users use, maybe CI should try to use `level=test` one more time

I remember that previously there was some issues and it has been disabled here https://github.com/ClickHouse/ClickHouse/pull/48424#issuecomment-1637722113 (cc @kssenii )

pamarcos replied:
I also gave it some thought and decided to add a new setting to enable this sort of traces just for the queries that are of our interest: [`955f537` (#71473)](https://github.com/ClickHouse/ClickHouse/pull/71473/commits/955f537bd5ef2f4a29717ac4999ce2af47b4c039)

So, for the stateless test it's enabled through `query_metric_log_debug` to help debugging it, but it doesn't print any noise during normal behavior. Does it work for you?

azat replied:
Yep, thank you!

pamarcos replied:
> Though settings just for this looks too much

Yep, completely agree. After giving it another thought, I decided to keep the logs for a while to ensure everything is good on the CI, but remove them before the release of the next version.
